src/render_tasks/d3d12_cubemap_convolution.hpp

```diff
@@ -0,0 +1,280 @@
+#pragma once
+
+#include "../d3d12/d3d12_renderer.hpp"
+#include "../d3d12/d3d12_defines.hpp"
+#include "../d3d12/d3d12_functions.hpp"
+#include "../d3d12/d3d12_constant_buffer_pool.hpp"
+#include "../d3d12/d3d12_structured_buffer_pool.hpp"
+#include "../d3d12/d3d12_model_pool.hpp"
+#include "../d3d12/d3d12_resource_pool_texture.hpp"
+#include "../frame_graph/frame_graph.hpp"
+#include "../scene_graph/camera_node.hpp"
+#include "../d3d12/d3d12_pipeline_registry.hpp"
+#include "../engine_registry.hpp"
+
+#include "../platform_independend_structs.hpp"
+#include "d3d12_imgui_render_task.hpp"
+#include "../scene_graph/camera_node.hpp"
+
+namespace wr
+{
+	struct CubemapConvolutionTaskData
+	{
+		D3D12Pipeline* in_pipeline;
+
+		TextureHandle in_radiance;
+		TextureHandle out_irradiance;
+
+		std::shared_ptr<ConstantBufferPool> camera_cb_pool;
+		D3D12ConstantBufferHandle* cb_handle;
+
+		DirectX::XMMATRIX proj_mat;
+		DirectX::XMMATRIX view_mat[6];
+
+		bool should_run = true;
+	};
+
+	namespace internal
+	{
+		struct ProjectionView_CBuffer
+		{
+			DirectX::XMMATRIX m_projection;
+			DirectX::XMMATRIX m_view[6];
+		};
+
+		inline void SetupCubemapConvolutionTask(RenderSystem& rs, FrameGraph& fg, RenderTaskHandle handle, TextureHandle in_radiance, TextureHandle out_irradiance)
+		{
+			auto& n_render_system = static_cast<D3D12RenderSystem&>(rs);
+			auto& data = fg.GetData<CubemapConvolutionTaskData>(handle);
+
+			auto& ps_registry = PipelineRegistry::Get();
+			data.in_pipeline = (D3D12Pipeline*)ps_registry.Find(pipelines::cubemap_convolution);
+
+			data.in_radiance = in_radiance;
+			data.out_irradiance = out_irradiance;
+
+			data.camera_cb_pool = rs.CreateConstantBufferPool(2);
+			data.cb_handle = static_cast<D3D12ConstantBufferHandle*>(data.camera_cb_pool->Create(sizeof(ProjectionView_CBuffer)));
+
+			data.proj_mat = DirectX::XMMatrixPerspectiveFovRH(DirectX::XMConvertToRadians(90.0f), 1.0f, 0.1f, 10.0f);
+
+			data.view_mat[0] = DirectX::XMMatrixLookAtRH(DirectX::XMVectorSet(0.0f, 0.0f, 0.0f, 1.0f),
```

Comment:
You can replace this with a loop. And only define the different forward vectors hardcoded.